src/workerd/io/io-context.h

```diff
@@ -901,200 +892,166 @@ kj::PromiseForResult<Func, Worker::Lock&> IoContext::run(
 
-  if constexpr (jsg::isVoid<T>()) {
-    auto [ jsPromise, resolver ] = js.newPromiseAndResolver<kj::Maybe<kj::Exception>>();
-
-    addTask(promise.then([]() -> kj::Maybe<kj::Exception> {
-      return kj::none;
-    }, [](kj::Exception&& exception) -> kj::Maybe<kj::Exception> {
-      return kj::mv(exception);
-    }).then([this, resolver = kj::mv(resolver), ilOrCs = kj::mv(ilOrCs),
-             maybeAsyncContext = jsg::AsyncContextFrame::currentRef(lock)]
-            (kj::Maybe<kj::Exception>&& maybeException) mutable {
-      return run([resolver = kj::mv(resolver),
-                  maybeException = kj::mv(maybeException),
-                  maybeAsyncContext = kj::mv(maybeAsyncContext)](Worker::Lock& lock) mutable {
-        jsg::AsyncContextFrame::Scope asyncScope(lock, maybeAsyncContext);
-        // We don't use `resolver.reject()` for exceptions here because if we convert the
-        // kj::Exception into a JS Error here, it won't have a useful stack trace. V8 can generate
-        // a good stack trace as long as we construct the Error inside of a promise continuation,
-        // so we use a `.then()` below that acutally checks for the kj::Exception and turn it into
-        // a JS Error.
-        resolver.resolve(kj::mv(maybeException));
-      }, kj::mv(ilOrCs));
-    }));
-
-    return jsPromise.then(js, [](jsg::Lock& js, kj::Maybe<kj::Exception>&& maybeEx) -> Result {
-      KJ_IF_SOME(e, maybeEx) {
-        // Now that we're in a promise continuation, we can convert the error and get a good stack
-        // trace.
-        js.throwException(kj::mv(e));
-      }
-    });
-  } else {
-    typedef kj::OneOf<T, kj::Exception> ResultOrException;
-    typedef kj::OneOf<Result, kj::Exception> IoResultOrException;
-    auto [ jsPromise, resolver ] = js.newPromiseAndResolver<IoResultOrException>();
-
-    addTask(promise.then([](T&& result) -> ResultOrException {
-      return kj::mv(result);
-    }, [](kj::Exception&& exception) -> ResultOrException {
-      return kj::mv(exception);
-    }).then([this, resolver = kj::mv(resolver), ilOrCs = kj::mv(ilOrCs),
-             maybeAsyncContext = jsg::AsyncContextFrame::currentRef(lock)]
-            (ResultOrException&& resultOrException) mutable {
-      return run([this, resolver = kj::mv(resolver),
-                  resultOrException = kj::mv(resultOrException),
-                  maybeAsyncContext = kj::mv(maybeAsyncContext)](Worker::Lock& lock) mutable {
-        jsg::AsyncContextFrame::Scope asyncScope(lock, maybeAsyncContext);
-        KJ_SWITCH_ONEOF(resultOrException) {
-          KJ_CASE_ONEOF(result, T) {
-            if constexpr (addIoOwn) {
-              resolver.resolve(addObject(kj::heap(kj::mv(result))));
-            } else {
-              (void)this; // used only in one branch of constexpr expression
-              resolver.resolve(kj::mv(result));
-            }
+  auto [ jsPromise, resolver ] = js.newPromiseAndResolver<ExceptionOr<Result>>();
```

Comment:
There is a possibility that `newPromiseAndResolver(...)` here could throw `JsExceptionThrown` before the `func` is attached to anything, Is that ok here? It should only ever happen in fatal cases, of course, but if part of the intent is to ensure that `func` outlives `promise`, then I think maybe this might be a condition where it may not?

Reply:
Technically at this point, both `func` and `promise` are on the stack, but since `func` was passed in by rvalue reference whereas `promise` was by value, `promise`'s destructor will run first. But I that is rather... precarious. I should make it more explicit.

Reply:
Decided to just comment on it instead of change it: 07f8cb10

Also moved `newPromiseAndResolver` down, though, since it wasn't needed yet.